### PR TITLE
WIP: First attempt at adding entrypoint to container config

### LINKED ISSUE #178: Allow a task or container to override an image's ENTRYPOINT
Say I'm using an image that someone else built, which has set `ENTRYPOINT someTool` in its Dockerfile. When using that image in batect, I can do `command: someCmd`, and inside the container it will run `someTool someCmd`. That's all expected behaviour.

Where it gets tricky is if I want to override the `ENTRYPOINT` that the image has. If I want to get an interactive shell in the container for debugging, or if I want to run two commands with `&&`, I don't think it's possible. When I set `command: sh`, it ends up running `someTool sh` in the container, which is not what I wanted.

When using `docker run` from the CLI, it's easy to fix - I can just set the `--entrypoint` flag to whatever I need. But AFAICT batect doesn't let me set the entrypoint on either a container or a task, which means the only way to override it is to create a new Dockerfile and build my own image with a different `ENTRYPOINT`.

Would you be interested in a PR that allows setting `entrypoint` on tasks or containers or both?

### PR DESCRIPTION
**Pull request by camjackson**
Opening this mostly to get feedback and see if I'm on the right track.

- Haven't added docs yet
- Haven't tested it manually yet to see if it actually works or not
- It was a pretty crude process - I essentially just did global searches for `command` and `workingDirectory` and tried to add `entryPoint` in wherever it looked like it belonged. Then I just fixed all the compiler errors and the 2 tests that I broke.

Resolves #178.

### CODE REVIEW

**Comment by camjackson** [comment]
Ok so I never ran the integration or journey tests before pushing. So those are broken. And my initial manual test is that it doesn't work either. It's parsing the yaml successfully but not actually overriding the entrypoint. Will come back to this at some point in the next few days.

File: app/src/main/kotlin/batect/docker/DockerContainerCreationRequestFactory.kt [hunk 1]
```diff
@@ -51,6 +52,7 @@ class DockerContainerCreationRequestFactory(
             image,
             network,
             if (command != null) command.parsedCommand else emptyList(),
+            if (entrypoint != null) entrypoint.parsedCommand else emptyList(),
```
Comment (charleskorn) [inline app/src/main/kotlin/batect/docker/DockerContainerCreationRequestFactory.kt]: We might have to do something slightly different here - empty list for an entrypoint means 'use system default' instead of 'use image default' (source: https://docs.docker.com/engine/api/v1.37/#operation/ContainerCreate)

File: app/src/main/kotlin/batect/execution/ContainerDependencyGraph.kt [hunk 2]
```diff
@@ -72,13 +72,15 @@ data class ContainerDependencyGraph(val config: Configuration, val task: Task, v
 
             val dependencyNodes = resolveDependencies(dependencies, nodesAlreadyCreated, newPath)
             val command = containerCommandResolver.resolveCommand(container, task)
+            val entrypoint = container.entrypoint
```
Comment (charleskorn) [inline app/src/main/kotlin/batect/execution/ContainerDependencyGraph.kt]: If you just want to add support for overriding the entrypoint on a container level (as opposed to at the task level) in this PR, you could get away with omitting this - anything that needs the entrypoint could take it from `container`. (The reason we have `command` here is so that we can store the command to use in the dependency graph node and so everything downstream of here doesn't have to know whether the command is coming from the task or the container.)

File: app/src/unitTest/kotlin/batect/docker/DockerAPISpec.kt [hunk 3]
```diff
@@ -121,7 +121,8 @@ object DockerAPISpec : Spek({
                 val image = DockerImage("the-image")
                 val network = DockerNetwork("the-network")
                 val command = listOf("doStuff")
-                val request = DockerContainerCreationRequest(image, network, command, "some-host", "some-host", emptyMap(), "/some-dir", emptySet(), emptySet(), HealthCheckConfig(), null, false, false, emptySet(), emptySet())
+                val entrypoint = listOf("sh")
+                val request = DockerContainerCreationRequest(image, network, command, entrypoint, "some-host", "some-host", emptyMap(), "/some-dir", emptySet(), emptySet(), HealthCheckConfig(), null, false, false, emptySet(), emptySet())
```
Comment (charleskorn) [inline app/src/unitTest/kotlin/batect/docker/DockerAPISpec.kt]: Not part of this PR, but I need to come back and do something about this, that parameter list is getting way out of hand...

**Comment by camjackson** [comment]
@charleskorn I ended up not needing it myself, so I sort of lost motivation to implement. Happy for you to take it over (or start fresh) if you're keen.
